paystack/src/main/java/co/paystack/android/TransactionManager.java

```diff
@@ -105,180 +87,171 @@ void chargeCard(Activity activity, Charge charge, Paystack.TransactionCallback t
         }
 
         this.activity = activity;
-        this.charge = charge;
         this.transactionCallback = transactionCallback;
-        this.transaction = new Transaction();
 
-        charge();
+        validateCardThenInitTransaction(publicKey, charge);
     }
 
-    private void charge() {
+    private void validateCardThenInitTransaction(String publicKey, Charge charge) {
         try {
             if (charge.getCard() == null || !charge.getCard().isValid()) {
                 final CardSingleton si = CardSingleton.getInstance();
                 synchronized (si) {
                     si.setCard(charge.getCard());
                 }
-                new CardAsyncTask().execute();
+                new CardAsyncTask(publicKey, charge).execute();
             } else {
-                initiate();
-                sendChargeToServer();
+                if (TransactionManager.PROCESSING) {
+                    throw new ProcessingException();
+                }
+
+                setProcessingOn();
+
+                String deviceId = "androidsdk_" + Settings.Secure.getString(activity.getContentResolver(), Settings.Secure.ANDROID_ID);
+                initiateTransaction(publicKey, charge, deviceId);
             }
         } catch (Exception ce) {
             Log.e(LOG_TAG, ce.getMessage(), ce);
             if (!(ce instanceof ProcessingException)) {
                 setProcessingOff();
             }
-            transactionCallback.onError(ce, transaction);
+            transactionCallback.onError(ce, NO_TRANSACTION);
         }
     }
 
-    private void sendChargeToServer() {
-        try {
-            initiateChargeOnServer();
-        } catch (Exception ce) {
-            Log.e(LOG_TAG, ce.getMessage(), ce);
-            notifyProcessingError(ce);
-        }
+    private void initiateTransaction(String publicKey, Charge charge, String deviceId) {
+        paystackRepository.initializeTransaction(publicKey, charge, deviceId, new ApiCallback<TransactionInitResponse>() {
```

Comment:
Out of interest, since you are writing a whole lot of new code here, would it not have helped you to place this logic in a Kotlin file and call the Kotlin code from Java? Just trying to help future Michael..

Reply:
I understand the appeal but we probably won't be supporting this much longer